File: lib/lexer.ml
open Char.Ascii

type t = {
  src: string;
  pos: int;
  line: int;
  llp: int;
  ch: char option;
}

let create src = {
  src = src;
  pos = 0;
  line = 0;
  llp = 0;
  ch = if String.length src = 0 then None else Some src.[0];
}

let is_start_of_id c = is_letter c || c = '_' || c = '$'

let is_part_of_id c = is_start_of_id c || is_digit c

let is_processable c =
  match c with
  | '.' | '!' | '*' | '/' | '+' | '-' | '^' | '&' | '|' | '(' | ')' | '[' | ']' | '{' | '}' | '"'
  | '\'' | '<' | '>' | '=' | '%' | '@' | ':' | ';' | '#' | ',' | '~' -> true
  | _ -> is_start_of_id c || is_digit c || Util.is_space c

let advance l =
  let pos = l.pos + 1 in
  let line, llp = match l.ch with
  | None -> l.line, l.llp
  | Some x -> if x = '\n' then l.line + 1, pos else l.line, l.llp in
  let ch = if String.length l.src > pos then Some l.src.[pos] else None in
  {
    src = l.src;
    pos = pos;
    line = line;
    llp = llp;
    ch = ch;
  }

let peek l x = if l.pos + x < String.length l.src then Some l.src.[l.pos + x] else None

let generate_token_and_advance l kind =
  let token : Token.t = {
    kind = kind;
    line = l.line;
    pos = l.pos - l.llp;
    length = 1;
    has_whitespace_after = Util.is_space @@ Option.value (peek l 1) ~default:' ';
  } in
  advance l, Some token

let rec seek l condition =
  if condition l.ch then
    let ch = String.make 1 @@ Option.value l.ch ~default:'?' in
    let l, remain = seek (advance l) condition in
    l, ch ^ remain
  else
    l, ""

let skip_whitespace l =
  let l, _ = seek l (fun x -> match x with
                     | None -> false
                     | Some c -> Util.is_space c) in
  l

let rec skip_c_comments l =
  match peek l 0, peek l 1 with
  | Some c1, Some c2 ->
      if c1 = '*' && c2 = '/' then advance l |> advance else advance l |> skip_c_comments
  | Some _, _ -> advance l |> skip_c_comments
  | _ -> l

let rec skip_cpp_comments l =
  match peek l 0 with
  | None -> l
  | Some c -> if c = '\n' then advance l else advance l |> skip_cpp_comments

let process_error l =
  let pos = l.pos in
  let l, value = seek l (fun x -> match x with
                         | None -> false
                         | Some c -> not @@ is_processable c) in
  let token : Token.t = {
    kind = Token.Error value;
    line = l.line;
    pos = pos - l.llp;
    length = l.pos - pos;
    has_whitespace_after = Util.is_space @@ Option.value l.ch ~default:' ';
  } in
  l, Some token

let process_id l =
  let pos = l.pos in
  let l, value = seek l (fun x -> match x with
                         | None -> false
                         | Some c -> is_part_of_id c) in
  let token : Token.t = {
    kind = Token.Identifier value;
    line = l.line;
    pos = pos - l.llp;
    length = l.pos - pos;
    has_whitespace_after = Util.is_space @@ Option.value l.ch ~default:' ';
  } in
  l, Some token

let rec process_str_seq l chr =
  match l.ch with
  | None -> l, ""
  | Some c1 ->
      if c1 = '\\' then
        match peek l 1 with
        | None -> advance l, "\\"
        | Some c2 ->
            if c2 = chr then
              let l, str = process_str_seq (advance l |> advance) chr in
              l, "\\" ^ String.make 1 chr ^ str
            else
              let l, str = process_str_seq (advance l) chr in
              l, "\\" ^ str
      else if c1 = chr then
        advance l, ""
      else
        let l, str = process_str_seq (advance l) chr in
        l, String.make 1 c1 ^ str

let process_string l =
  let pos, line, llp = l.pos, l.line, l.llp in
  let l, value = process_str_seq (advance l) '"' in
  let token : Token.t = {
    kind = Token.String value;
    line = line;
    pos = pos - llp;
    length = l.pos - pos;
    has_whitespace_after = Util.is_space @@ Option.value l.ch ~default:' ';
  } in
  l, Some token

let process_symbol l =
  let pos, line, llp = l.pos, l.line, l.llp in
  let l, value = process_str_seq (advance l) '\'' in
  let token : Token.t = {
    kind = Token.Symbol value;
    line = line;
    pos = pos - llp;
    length = l.pos - pos;
    has_whitespace_after = Util.is_space @@ Option.value l.ch ~default:' ';
  } in
  l, Some token

let process_num l =
  let _hex l =
    let pos = l.pos in
    let l = advance l |> advance in
    let l, value = seek l (fun x -> match x with
                           | None -> false
                           | Some c -> is_hex_digit c || c = '_' || c = '\'') in
    let x = ref 0 in
    for i = 0 to String.length value - 1 do
      let c = value.[i] in
      match c with
      | '\'' | '_' -> ()
      | _ ->
          let ascii_val = int_of_char c in
          if c >= 'a' && c <= 'f' then
            x := (!x lsl 4) lor (ascii_val - int_of_char 'a' + 0xA)
          else if c >= 'A' && c <= 'F' then
            x := (!x lsl 4) lor (ascii_val - int_of_char 'A' + 0xA)
          else
            x := (!x lsl 4) lor (ascii_val - int_of_char '0')
    done;
    let token : Token.t = {
      kind = Token.NumberHex !x;
      line = l.line;
      pos = pos - l.llp;
      length = l.pos - pos;
      has_whitespace_after = Util.is_space @@ Option.value l.ch ~default:' ';
    } in
    l, Some token in

  let _bin l =
    let pos = l.pos in
    let l = advance l |> advance in
    let l, value = seek l (fun x -> match x with
                           | None -> false
                           | Some c -> c = '0' || c = '1' || c = '_' || c = '\'') in
    let x = ref 0 in
    for i = 0 to String.length value - 1 do
      let c = value.[i] in
      match c with
      | '\'' | '_' -> ()
      | _ -> x := (!x lsl 1) lor (int_of_char c - int_of_char '0')
    done;
    let token : Token.t = {
      kind = Token.NumberBin !x;
      line = l.line;
      pos = pos - l.llp;
      length = l.pos - pos;
      has_whitespace_after = Util.is_space @@ Option.value l.ch ~default:' ';
    } in
    l, Some token in

  let _oct l =
    let pos = l.pos in
    let l = advance l in
    let l, value = seek l (fun x -> match x with
                           | None -> false
                           | Some c -> (c >= '0' && c <= '7') || c = '_' || c = '\'') in
    let x = ref 0 in
    for i = 0 to String.length value - 1 do
      let c = value.[i] in
      match c with
      | '\'' | '_' -> ()
      | _ -> x := (!x lsl 3) lor (int_of_char c - int_of_char '0')
    done;
    let token : Token.t = {
      kind = Token.NumberOct !x;
      line = l.line;
      pos = pos - l.llp;
      length = l.pos - pos;
      has_whitespace_after = Util.is_space @@ Option.value l.ch ~default:' ';
    } in
    l, Some token in

  let _float_dec l =
    let pos = l.pos in
    let l, value, has_period =
      let rec _get_str l has_period =
        match l.ch with
        | None -> l, "", has_period
        | Some c ->
            match c with
            | '\'' | '_' -> _get_str (advance l) has_period
            | '.' ->
                if has_period then l, "", true
                else
                  let l, value, _ = _get_str (advance l) true in
                  l, "." ^ value, true
            | _ ->
                if is_digit c then
                  let l, value, has_period = _get_str (advance l) has_period in
                  l, String.make 1 c ^ value, has_period
                else
                  l, "", has_period in
      _get_str l false in
    let token : Token.t = {
      kind = if has_period then Token.NumberFloat (float_of_string value)
                           else Token.Number (int_of_string value);
      line = l.line;
      pos = pos - l.llp;
      length = l.pos - pos;
      has_whitespace_after = Util.is_space @@ Option.value l.ch ~default:' ';
    } in
    l, Some token in

  if Option.get l.ch = '0' then
    match peek l 1 with
    | None ->
        let token : Token.t = {
          kind = Token.Number 0;
          line = l.line;
          pos = l.pos - l.llp;
          length = 1;
          has_whitespace_after = true;
        } in
        advance l, Some token
    | Some c ->
        match c with
        | 'x' -> _hex l
        | 'b' -> _bin l
        | '.' | '\'' | '_' -> _float_dec l
        | _ ->
            if is_digit c then _oct l
            else
              let token : Token.t = {
                kind = Token.Number 0;
                line = l.line;
                pos = l.pos - l.llp;
                length = 1;
                has_whitespace_after = true;
              } in
              advance l, Some token
  else
    _float_dec l

let next_token l =
  let l = skip_whitespace l in
  match l.ch with
  | None ->
      let eof_token : Token.t = {
        kind = Token.Eof;
        line = l.line;
        pos = l.pos;
        length = 1;
        has_whitespace_after = true;
      } in
      l, Some eof_token
  | Some c ->
      if not @@ is_processable c then
        process_error l
      else if is_start_of_id c then
        process_id l
      else if is_digit c then
        process_num l
      else match c with
      | '/' ->
          (match peek l 1 with
          | None -> generate_token_and_advance l Token.Slash
          | Some x ->
              match x with
              | '/' -> skip_cpp_comments l, None
              | '*' -> skip_c_comments l, None
              | _ -> generate_token_and_advance l Token.Slash)
      | '"' -> process_string l
      | '\'' -> process_symbol l
      | '(' -> generate_token_and_advance l Token.LParen
      | ')' -> generate_token_and_advance l Token.RParen
      | '[' -> generate_token_and_advance l Token.LBrack
      | ']' -> generate_token_and_advance l Token.RBrack
      | '{' -> generate_token_and_advance l Token.LCBrack
      | '}' -> generate_token_and_advance l Token.RCBrack
      | '<' -> generate_token_and_advance l Token.LArrow
      | '>' -> generate_token_and_advance l Token.RArrow
      | '+' -> generate_token_and_advance l Token.Plus
      | '-' -> generate_token_and_advance l Token.Minus
      | '*' -> generate_token_and_advance l Token.Asterisk
      | '%' -> generate_token_and_advance l Token.Percent
      | '&' -> generate_token_and_advance l Token.Ampersand
      | '|' -> generate_token_and_advance l Token.Pipe
      | '~' -> generate_token_and_advance l Token.Tilde
      | '^' -> generate_token_and_advance l Token.Circumflex
      | '=' -> generate_token_and_advance l Token.Eq
      | '!' -> generate_token_and_advance l Token.ExclMark
      | '@' -> generate_token_and_advance l Token.At
      | ':' -> generate_token_and_advance l Token.Colon
      | ';' -> generate_token_and_advance l Token.Semicolon
      | '.' ->
          (match peek l 1 with
          | None -> generate_token_and_advance l Token.Period
          | Some x ->
              if is_digit x then process_num l
              else generate_token_and_advance l Token.Period)
      | ',' -> generate_token_and_advance l Token.Comma
      | '#' -> generate_token_and_advance l Token.Hash
      | _ -> l, None

let rec tokenize l =
  let l, token = next_token l in
  match token with
  | None -> tokenize l
  | Some x ->
      match x.kind with
      | Token.Eof -> l, [], []
      | Token.Error _ ->
          let l, tokens, errors = tokenize l in
          l, tokens, x :: errors
      | _ ->
          let l, tokens, errors = tokenize l in
          l, x :: tokens, errors

File: libalc/lexer.c
#include "alc/lexer.h"
#include "alc/defs.h"
#include "alc/token.h"
#include "allocs/alloc_arena.h"
#include "containers/vector.h"
#include "global.h"
#include <ctype.h>
#include <string.h>

#define CHAR_CASES             \
  CHAR_CASE_X('.', PERIOD)     \
  CHAR_CASE_X(',', COMMA)      \
  CHAR_CASE_X('!', EXCLMARK)   \
  CHAR_CASE_X('*', ASTERISK)   \
  CHAR_CASE_X('+', PLUS)       \
  CHAR_CASE_X('-', MINUS)      \
  CHAR_CASE_X('^', CIRCUMFLEX) \
  CHAR_CASE_X('&', AMPERSAND)  \
  CHAR_CASE_X('|', PIPE)       \
  CHAR_CASE_X('(', LPAREN)     \
  CHAR_CASE_X(')', RPAREN)     \
  CHAR_CASE_X('[', LBRACK)     \
  CHAR_CASE_X(']', RBRACK)     \
  CHAR_CASE_X('{', LCBRACK)    \
  CHAR_CASE_X('}', RCBRACK)    \
  CHAR_CASE_X('<', LARROW)     \
  CHAR_CASE_X('>', RARROW)     \
  CHAR_CASE_X('=', EQ)         \
  CHAR_CASE_X('%', PERCENT)    \
  CHAR_CASE_X('@', AT)         \
  CHAR_CASE_X(':', COLON)      \
  CHAR_CASE_X(';', SEMICOLON)  \
  CHAR_CASE_X('#', HASH)       \
  CHAR_CASE_X('~', TILDE)

static alc_token_t gen_token(const char *value, alc_token_type_t type, usize line, usize llp,
                             usize raw_pos, usize len, b8 has_whitespace_after);
static inline alc_token_t process_error(alc_lexer_t *l);
static inline alc_token_t process_string(alc_lexer_t *l);
static inline alc_token_t process_symbol(alc_lexer_t *l);
static inline alc_token_t process_id(alc_lexer_t *l);
static inline alc_token_t process_num(alc_lexer_t *l);
static inline alc_token_t process_num_hex(alc_lexer_t *l);
static inline alc_token_t process_num_bin(alc_lexer_t *l);
static inline alc_token_t process_num_oct(alc_lexer_t *l);
static inline b8 is_start_of_id(char c);
static inline b8 is_part_of_id(char c);
static inline b8 is_processable(char c);
static inline void skip_whitespace(alc_lexer_t *l);
static inline void skip_c_comments(alc_lexer_t *l);
static inline void skip_cpp_comments(alc_lexer_t *l);
static inline b8 is_whitespace_at(const alc_lexer_t *l, usize pos);
static inline char peek(const alc_lexer_t *l, usize adv);
static char *memcpy_cond(char *dst, const char *src, usize n, b8 (*cond)(char c));

static b8 skip_separators(char c)
{
  return c != '\'' && c != '_';
}

alc_lexer_t alc_lexer_create(const char *src)
{
  ALC_ASSERT(src != nullptr);

  return (alc_lexer_t){
    .src = src,
    .src_len = strlen(src),
  };
}

b8 alc_lexer_tokenize(alc_lexer_t *lexer, alc_token_t **out_tokens, usize *out_n)
{
  ALC_ASSERT(lexer != nullptr);
  ALC_ASSERT(out_tokens != nullptr);
  ALC_ASSERT(out_n != nullptr);

  alc_token_t *tokens = vector_create(alc_token_t);
  b8 failed = false;

  while (lexer->pos < lexer->src_len) {
    skip_whitespace(lexer);

    if ALC_UNLIKELY (lexer->pos >= lexer->src_len)
      break;

    char c = lexer->src[lexer->pos];
    switch (c) {
#define CHAR_CASE_X(_char, _type_name)                                                            \
  case (_char): {                                                                                 \
    if ALC_LIKELY (!failed) {                                                                     \
      vector_push(tokens,                                                                         \
                  gen_token(nullptr, ALC_TOKEN_TYPE_FULL_NAME(_type_name), lexer->line,           \
                            lexer->llp, lexer->pos, 1, is_whitespace_at(lexer, lexer->pos + 1))); \
    }                                                                                             \
    lexer->pos++;                                                                                 \
  } break;
      CHAR_CASES
#undef CHAR_CASE_X

    case '/': {
      char peeked = peek(lexer, 1);
      switch (peeked) {
      case '/': {
        skip_cpp_comments(lexer);
      } break;
      case '*': {
        skip_c_comments(lexer);
      } break;
      default:
        if ALC_LIKELY (!failed) {
          vector_push(tokens, gen_token(nullptr, ALC_TOKEN_TYPE_SLASH, lexer->line, lexer->llp,
                                        lexer->pos, 1, is_whitespace_at(lexer, lexer->pos + 1)));
        }
        lexer->pos++;
        break;
      }
    } break;

    case '"': {
      alc_token_t string_token = process_string(lexer);
      if ALC_LIKELY (!failed)
        vector_push(tokens, string_token);
      break;
    }
    case '\'': {
      alc_token_t symbol_token = process_symbol(lexer);
      if ALC_LIKELY (!failed)
        vector_push(tokens, symbol_token);
      break;
    }
    default:
      if (is_start_of_id(c)) {
        alc_token_t id_token = process_id(lexer);
        if ALC_LIKELY (!failed)
          vector_push(tokens, id_token);
        break;
      } else if (isdigit(c)) {
        alc_token_t num_token = process_num(lexer);
        if ALC_LIKELY (!failed)
          vector_push(tokens, num_token);
        break;
      }

      if ALC_LIKELY (!failed) {
        failed = true;
        vector_clear(tokens);
      }
      printf("FAILURE: %zu:%zu: %c\n", lexer->line + 1, lexer->pos - lexer->llp,
             lexer->src[lexer->pos]);
      vector_push(tokens, process_error(lexer));
      break;
    }
  }

  if ALC_LIKELY (vector_get_length(tokens) > 0)
    *out_tokens = vector_to_array(tokens, out_n);
  else {
    *out_tokens = nullptr;
    *out_n = 0;
  }
  vector_destroy(tokens);
  return !failed;
}

static alc_token_t gen_token(const char *value, alc_token_type_t type, usize line, usize llp,
                             usize raw_pos, usize len, b8 has_whitespace_after)
{
  char *value_in_arena = nullptr;
  if (value != nullptr) {
    usize val_len = strlen(value);
    value_in_arena = alloc_arena_allocate_aligned(&ctx()->arena, val_len + 1, 1);
    memcpy(value_in_arena, value, val_len + 1);
  }

  return (alc_token_t){
    .line = line,
    .pos = raw_pos - llp,
    .len = len,
    .value = value_in_arena,
    .type = type,
    .has_whitespace_after = has_whitespace_after,
  };
}

static inline alc_token_t process_error(alc_lexer_t *l)
{
  usize start = l->pos;
  for (; l->pos < l->src_len && !is_processable(l->src[l->pos]); l->pos++)
    ;

  usize len = l->pos - start;
  const char *s_ptr = &l->src[start];
  char *value = alloc_arena_allocate_aligned(&ctx()->arena, sizeof(char) * (len + 1), 1);
  memcpy(value, s_ptr, len * sizeof(char));
  value[len] = 0;

  alc_token_t token = gen_token(value, ALC_TOKEN_TYPE_ERROR, l->line, l->llp, start, len,
                                is_whitespace_at(l, l->pos));
  return token;
}

static char *get_string_value(alc_lexer_t *l, char terminator)
{
  l->pos++;

  usize start = l->pos;
  while (l->pos < l->src_len) {
    char c = l->src[l->pos];
    if (c == terminator) {
      l->pos++;
      break;
    } else if (c == '\n' || c == '\r') {
      l->pos++;
      l->line++;
      l->llp = l->pos;
      continue;
    } else if (c == '\\' && peek(l, 1) == terminator) {
      l->pos += 2;
      continue;
    }
    l->pos++;
  }

  usize len = l->pos - start;
  char *data = alloc_arena_allocate_aligned(&ctx()->arena, len, 1);
  memcpy(data, &l->src[start], len - 1);
  data[len - 1] = 0;

  return data;
}

static inline alc_token_t process_string(alc_lexer_t *l)
{
  usize start = l->pos, line = l->line, llp = l->llp;
  char *value = get_string_value(l, '\"');
  return gen_token(value, ALC_TOKEN_TYPE_STRING, line, llp, start, l->pos - start,
                   is_whitespace_at(l, l->pos));
}

static inline alc_token_t process_symbol(alc_lexer_t *l)
{
  usize start = l->pos, line = l->line, llp = l->llp;
  char *value = get_string_value(l, '\'');
  return gen_token(value, ALC_TOKEN_TYPE_SYMBOL, line, llp, start, l->pos - start,
                   is_whitespace_at(l, l->pos));
}

static inline alc_token_t process_id(alc_lexer_t *l)
{
  usize start = l->pos;
  for (; l->pos < l->src_len && is_part_of_id(l->src[l->pos]); l->pos++)
    ;
  usize len = l->pos - start;
  char *value = alloc_arena_allocate_aligned(&ctx()->arena, len + 1, 1);
  memcpy(value, &l->src[start], len);
  value[len] = 0;

  return gen_token(value, ALC_TOKEN_TYPE_ID, l->line, l->llp, start, len,
                   is_whitespace_at(l, l->pos));
}

static inline alc_token_t process_num(alc_lexer_t *l)
{
  if (l->src[l->pos] == '0') {
    switch (peek(l, 1)) {
    case 'x':
      return process_num_hex(l);
    case 'b':
      return process_num_bin(l);
    case 'o':
      return process_num_oct(l);
    default:
      break;
    }
  }

  b8 is_float = false;
  usize start = l->pos;
  for (; l->pos < l->src_len && (isdigit(l->src[l->pos]) || l->src[l->pos] == '\'' ||
                                 l->src[l->pos] == '_' || l->src[l->pos] == '.');
       l->pos++) {
    if (l->src[l->pos] == '.') {
      if ALC_UNLIKELY (is_float)
        break;
      is_float = true;
    }
  }

  usize len = l->pos - start;
  char *value = alloc_arena_allocate_aligned(&ctx()->arena, len + 1, 1);
  char *end = memcpy_cond(value, &l->src[start], len, skip_separators);
  *end = 0;
  return gen_token(value, is_float ? ALC_TOKEN_TYPE_NUMBER_FLOAT : ALC_TOKEN_TYPE_NUMBER, l->line,
                   l->llp, start, len, is_whitespace_at(l, l->pos));
}

static inline alc_token_t process_num_hex(alc_lexer_t *l)
{
  l->pos += 2;
  usize start = l->pos;
  for (; l->pos < l->src_len &&
         (isxdigit(l->src[l->pos]) || l->src[l->pos] == '\'' || l->src[l->pos] == '_');
       l->pos++)
    ;
  usize len = l->pos - start;
  char *value = alloc_arena_allocate_aligned(&ctx()->arena, len + 1, 1);
  char *end = memcpy_cond(value, &l->src[start], len, skip_separators);
  *end = 0;

  return gen_token(value, ALC_TOKEN_TYPE_NUMBER_HEX, l->line, l->llp, start - 2, len + 2,
                   is_whitespace_at(l, l->pos));
}

static inline alc_token_t process_num_bin(alc_lexer_t *l)
{
  l->pos += 2;
  usize start = l->pos;
  for (; l->pos < l->src_len && (l->src[l->pos] == '0' || l->src[l->pos] == '1' ||
                                 l->src[l->pos] == '\'' || l->src[l->pos] == '_');
       l->pos++)
    ;
  usize len = l->pos - start;
  char *value = alloc_arena_allocate_aligned(&ctx()->arena, len + 1, 1);
  char *end = memcpy_cond(value, &l->src[start], len, skip_separators);
  *end = 0;

  return gen_token(value, ALC_TOKEN_TYPE_NUMBER_BIN, l->line, l->llp, start - 2, len + 2,
                   is_whitespace_at(l, l->pos));
}

static inline alc_token_t process_num_oct(alc_lexer_t *l)
{
  l->pos += 2;
  usize start = l->pos;
  for (; l->pos < l->src_len && ((l->src[l->pos] >= '0' && l->src[l->pos] <= '7') ||
                                 l->src[l->pos] == '\'' || l->src[l->pos] == '_');
       l->pos++)
    ;
  usize len = l->pos - start;
  char *value = alloc_arena_allocate_aligned(&ctx()->arena, len + 1, 1);
  char *end = memcpy_cond(value, &l->src[start], len, skip_separators);
  *end = 0;

  return gen_token(value, ALC_TOKEN_TYPE_NUMBER_OCT, l->line, l->llp, start - 2, len + 2,
                   is_whitespace_at(l, l->pos));
}

static inline b8 is_start_of_id(char c)
{
  return isalpha(c) || c == '_' || c == '$';
}

static inline b8 is_part_of_id(char c)
{
  return is_start_of_id(c) || isdigit(c);
}

static inline b8 is_processable(char c)
{
  switch (c) {
#define CHAR_CASE_X(_char, _type_name) case (_char):
    CHAR_CASES
#undef CHAR_CASE_X
  case '"':
  case '\'':
    return true;
  default:
    return is_start_of_id(c) || isdigit(c) || isspace(c);
  }
}

static inline void skip_whitespace(alc_lexer_t *l)
{
  for (; l->pos < l->src_len && isspace(l->src[l->pos]); l->pos++) {
    switch (l->src[l->pos]) {
    case '\n':
    case '\r':
      l->llp = l->pos + 1;
      l->line++;
      break;
    default:
      break;
    }
  }
}

static inline void skip_c_comments(alc_lexer_t *l)
{
  for (; l->pos < l->src_len; l->pos++) {
    char c1 = peek(l, 0), c2 = peek(l, 1);
    if (c1 == '*' && c2 == '/') {
      l->pos += 2;
      break;
    }

    if (c1 == '\n' || c1 == '\r') {
      l->line++;
      l->llp = l->pos + 1;
    }
  }
}

static inline void skip_cpp_comments(alc_lexer_t *l)
{
  for (; l->pos < l->src_len; l->pos++) {
    char c = l->src[l->pos];
    if (c == '\n' || c == '\r') {
      l->pos++;
      l->line++;
      l->llp = l->pos;
      break;
    }
  }
}

static inline b8 is_whitespace_at(const alc_lexer_t *l, usize pos)
{
  return pos >= l->src_len || isspace(l->src[pos]);
}

static inline char peek(const alc_lexer_t *l, usize adv)
{
  return l->pos + adv >= l->src_len ? 0 : l->src[l->pos + adv];
}

static char *memcpy_cond(char *dst, const char *src, usize n, b8 (*cond)(char c))
{
  for (; n; n--, src++) {
    if (cond(*src))
      *dst++ = *src;
  }
  return dst;
}
